iocore/hostdb/P_RefCountCache.h

```diff
@@ -0,0 +1,896 @@
+  RefCountCacheHashEntry() : expiry_entry(NULL), meta(0, 0){};
+
+  // make these values comparable -- so we can sort them
+  bool
+  operator<(const RefCountCacheHashEntry &v2) const
+  {
+    return this->meta.expiry_time < v2.meta.expiry_time;
+  };
+};
+// Since the hashing values are all fixed size, we can simply use a classAllocator to avoid mallocs
+extern ClassAllocator<RefCountCacheHashEntry> refCountCacheHashingValueAllocator;
+extern ClassAllocator<PriorityQueueEntry<RefCountCacheHashEntry *>> expiryQueueEntry;
+
+struct RefCountCacheHashing {
+  typedef uint64_t ID;
+  typedef uint64_t const Key;
+  typedef RefCountCacheHashEntry Value;
+  typedef DList(RefCountCacheHashEntry, item_link) ListHead;
+
+  static ID
+  hash(Key key)
+  {
+    return key;
+  }
+  static Key
+  key(Value const *value)
+  {
+    return value->meta.key;
+  }
+  static bool
+  equal(Key lhs, Key rhs)
+  {
+    return lhs == rhs;
+  }
+};
+
+// The RefCountCachePartition is simply a map of key -> Ptr<YourClass>
+// We partition the cache to reduce lock contention
+template <class C> class RefCountCachePartition
+{
+public:
+  RefCountCachePartition(unsigned int part_num, uint64_t max_size, unsigned int max_items, RecRawStatBlock *rsb = NULL);
+  Ptr<C> get(uint64_t key);
+  void put(uint64_t key, C *item, int size = 0, int expire_time = 0);
+  void erase(uint64_t key, int expiry_time = -1);
+
+  void clear();
+  bool is_full();
+  bool reserve(unsigned int);
+
+  const size_t count();
+  void copy(Vec<RefCountCacheHashEntry *> &items);
+
+  typedef typename TSHashTable<RefCountCacheHashing>::iterator iterator_type;
+  typedef typename TSHashTable<RefCountCacheHashing>::self hash_type;
+  typedef typename TSHashTable<RefCountCacheHashing>::Location location_type;
+  TSHashTable<RefCountCacheHashing> *get_map();
+
+  Ptr<ProxyMutex> lock; // Lock
+
+private:
+  void metric_inc(RefCountCache_Stats metric_enum, int64_t data);
+
+  unsigned int part_num;
+
+  uint64_t max_size;
+  unsigned int max_items;
+
+  uint64_t size;
+  unsigned int items;
+
+  hash_type item_map;
+
+  PriorityQueue<RefCountCacheHashEntry *> expiry_queue;
+  RecRawStatBlock *rsb;
+};
+
+template <class C>
+RefCountCachePartition<C>::RefCountCachePartition(unsigned int part_num, uint64_t max_size, unsigned int max_items,
+                                                  RecRawStatBlock *rsb)
+{
+  this->part_num  = part_num;
+  this->max_size  = max_size;
+  this->max_items = max_items;
+  this->size      = 0;
+  this->items     = 0;
+  this->rsb       = rsb;
+
+  // Initialize lock
+  this->lock = new_ProxyMutex();
+}
+
+template <class C>
+Ptr<C>
+RefCountCachePartition<C>::get(uint64_t key)
+{
+  this->metric_inc(refcountcache_total_lookups_stat, 1);
+  location_type l = this->item_map.find(key);
+  if (l.isValid()) {
+    // found
+    this->metric_inc(refcountcache_total_hits_stat, 1);
+    return make_ptr((C *)l.m_value->item.get());
```

Comment:
Oh I see, you are converting from ``Ptr<RefCountObj>`` to ``Ptr<C>`` where ``C`` inherits from ``RefCountObj``. I guess that this would have to be a ``dynamic_cast`` then :-/